build-system/common/transform-cache.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const crypto = require('crypto');
+const fs = require('fs-extra');
+const path = require('path');
+
+/**
+ * Cache for storing transformed files on both memory and on disk.
+ */
+class TransformCache {
+  constructor(cacheDir) {
+    /** @type {string} */
+    this.cacheDir = cacheDir;
+    fs.ensureDirSync(cacheDir);
+
+    /** @type {Map<string, Promise<Buffer|string>>} */
+    this.transformMap = new Map();
+
+    /** @type {Set<string>} */
+    this.fsCache = new Set(fs.readdirSync(cacheDir));
+  }
+
+  /**
+   * @param {string} hash
+   * @return {null|Promise<Buffer|string>}
+   */
+  get(hash) {
+    const cached = this.transformMap.get(hash);
+    if (cached) {
+      return cached;
+    }
+    if (this.fsCache.has(hash)) {
+      const transformedPromise = fs.readFile(path.join(this.cacheDir, hash));
+      this.transformMap.set(hash, transformedPromise);
+      return transformedPromise;
+    }
+    return null;
+  }
+
+  /**
+   * @param {string} hash
+   * @param {Promise<string>} transformPromise
+   */
+  set(hash, transformPromise) {
+    if (this.transformMap.has(hash)) {
+      throw new Error(
+        `Read race occured. Attempting to transform a file twice.`
+      );
+    }
+
+    this.transformMap.set(hash, transformPromise);
+    transformPromise.then((contents) => {
+      fs.outputFile(path.join(this.cacheDir, hash), contents);
```

Comment:
Nit: might be nice to include a file extension?

Reply:
WDYM? Previously we had `[hash].json`. Could make it `[hash].cache`?

Reply:
We're writing the raw files now, so `.js` or `.css`?